### ISSUE TITLE
error when use chrome  54.0.2840.71 (64-bit)

### ISSUE THREAD

**Issue by fix2015** [web-push-libs/web-push#254]
Hi, try use your library, all did like you wrote, but on Firefox it work perfect with chrome 54.0.2840.71 (64-bit) i get error

` [WebPushError: Received unexpected response code]
  name: 'WebPushError',
  message: 'Received unexpected response code',
  statusCode: 400,
  headers: 
   { 'content-type': 'text/html; charset=UTF-8',
     date: 'Wed, 09 Nov 2016 13:38:11 GMT',
     expires: 'Wed, 09 Nov 2016 13:38:11 GMT',
     'cache-control': 'private, max-age=0',
     'x-content-type-options': 'nosniff',
     'x-frame-options': 'SAMEORIGIN',
     'x-xss-protection': '1; mode=block',
     server: 'GSE',
     'alt-svc': 'quic=":443"; ma=2592000; v="36,35,34"',
     'accept-ranges': 'none',
     vary: 'Accept-Encoding',
     connection: 'close' },
  body: '<HTML>\n<HEAD>\n<TITLE>UnauthorizedRegistration</TITLE>\n</HEAD>\n<BODY BGCOLOR="#FFFFFF" TEXT="#000000">\n<H1>UnauthorizedRegistration</H1>\n<H2>Error 400</H2>\n</BODY>\n</HTML>\n' `
Maybe you know where can be pb?

**Comment by fix2015** [comment]
with Firefox i get this and all work 
`{ statusCode: 201,
  body: '',
  headers: 
   { 'access-control-allow-headers': 'content-encoding,encryption,crypto-key,ttl,encryption-key,content-type,authorization',
     'access-control-allow-methods': 'POST',
     'access-control-allow-origin': '*',
     'access-control-expose-headers': 'location,www-authenticate',
     'content-type': 'text/html; charset=UTF-8',
     date: 'Wed, 09 Nov 2016 13:38:23 GMT',
     location: 'https://updates.push.services.mozilla.com/m/gAAAAABYIybPHQEgecmFA8tmYFj7W8zFMhgTrzwYv5AJgTDFnhxXjhfKANlCVPCvo0WmKA6cXKglWa6X7jD1J7IRYPlvhaMhR1yDssoDd9WckMu-xzuKhYlmr1quOA8Ooz2NaHLoLIvKn30i789IkaRQ3tYQZn3wEqhEf2nwX61DfCqvajzr9-V-PCaAulA5fA8YRk9Ooxzd',
     server: 'cyclone/1.1',
     ttl: '2419200',
     'content-length': '0',
     connection: 'Close' } }
`

**Comment by fix2015** [comment]
`/**
 * Created by semianchuk on 04.04.16.
 */
var path = require('path');
var express = require('express');
var bodyParser = require('body-parser');

var app = express().use(express.static(
    path.join(__dirname, '/app/')
))

app.use(bodyParser.json()); // for parsing application/json
app.use(bodyParser.urlencoded({ extended: true })); // for parsing application/x-www-form-urlencoded

const webpush = require('web-push');

app.get('/send', function(req, res){
    console.log('start')

// VAPID keys should only be generated only once.
    const vapidKeys = webpush.generateVAPIDKeys();

    webpush.setGCMAPIKey('*******');
    webpush.setVapidDetails(
        'mailto:******,
        vapidKeys.publicKey,
        vapidKeys.privateKey
    );

// This is the same output of calling JSON.stringify on a PushSubscription
    const pushSubscription = {"endpoint":"https://fcm.googleapis.com/fcm/send/dG5e3_E9xcA:APA91bERn9EvUgrV9qc8UJ54v_dFpL4vY_xEcSPoLbJec6G1buVoaRIaybdY9eY4VlZrlh089-la3afTrCyOXLSzpozmEviMGnF02zvMxWmv-rT6PB7sRrtHd5OxphMhuoAYjTvOuL_f","keys":{"p256dh":"BPjvrIqi6b-fLdvBWdrW-OMqQtvgELG1YdR7AceTxry1XvHkcJUmbn2CaE4nSaBa-0x2zLWXtAd9JxhxQXABvik=","auth":"kYcUxDAMLWO8rPDLgxXFFg=="}}
/*    webpush.sendNotification(pushSubscription, 'Your Push Payload Text').then(function(data){
        console.log(data)
    });*/
    webpush.sendNotification(pushSubscription.endpoint, {
        TTL: 200,
    });

});


app.get('/sendNotification', function(req, res) {
    //const pushSubscription = {"endpoint":"https://fcm.googleapis.com/fcm/send/e2L42wAB2PQ:APA91bF3cidItVPAqeLZwgWWkMSoeWPbE2xQ05xSsyf19ZpSOUKTrwkfsIcnr3NvQxV0aUpOH1WJD1UFRSqJ8k_4pyaI-ZtuGFQOeLiWTPV-RwodaQcL-BimSDwg8hyRTZlKnKTlqm_n","keys":{"p256dh":"BAHkHv8wYJ_GpfhT2ubgb2vEm_pT_LE2A2X0RLZhf35ixHvLdwwDXpI_tbeUov4UpDBEyLigHauzxz6DuL9GKns=","auth":"BvKuhDlevQkDXvYZFwAetQ=="}}
    const pushSubscription = {"endpoint":"https://updates.push.services.mozilla.com/wpush/v1/gAAAAABYIyRLXevacQhqrx_I4Ua70M4T-22M_8LhH_yhYpyPay55q0W9O8ZHMR1KDDclaOM8gBuCrOhxTiyj4mouKdRTbYN5GqKGc5zkIN9x-nnxY7CRRk2sk4Az3MOP9PHSgU0dAdYV","keys":{"auth":"TSk-nMCrujjKTEXAONef8A","p256dh":"BFQkLFrNsPa4tjDACWsmRkmJOtEBo74YuPBZEWTdpS5_njLuaINsD0-yhpJ9Zb1hItBbTD-64fwjGO8MP5mebfg"}}
    console.log(pushSubscription.endpoint)
    webpush.sendNotification(pushSubscription, 'Your Push Payload Text').then(function (res) {
        console.log(res)
    })
    .catch(function(error){
        console.log(error);
        //process.exit(1);
    });

    console.log('done')
}, function(err) {
    console.log(err);
});

app.get('/', function(req, res){
    res.sendFile(__dirname + '/app/index.html');
});
console.log('Server running: http://localhost:8080')
app.listen(8080);
`

**Comment by gauntface** [comment]
What is the error that you are seeing?

**Comment by fix2015** [comment]
<HTML>
<HEAD>
<TITLE>UnauthorizedRegistration</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" TEXT="#000000">
<H1>UnauthorizedRegistration</H1>
<H2>Error 400</H2>
</BODY>
</HTML>

2016-11-09 17:03 GMT+02:00 Matt Gaunt <notifications@github.com>:

**Comment by p-nowak** [comment]
I have similar problem. On 3 different mac machines, sending notifications to Google and Firefox is working great, but using exactly the same data (encryption keys and user notification),  I can't make it working on linux server, because of this error - also tried with different machines

**Comment by p-nowak** [comment]
Unfortunately, that's not a platform problem, because on local Linux VM sending to Google is working properly

**Comment by vkammerer** [comment]
I have been seeing the same erratic behaviour with somehow seems to come from the FCM service. I have posted a comment on https://github.com/web-push-libs/web-push/issues/237

**Comment by spock123** [comment]
Same issue here - for what it's worth.

The strange thing is that it works beautifully when I'm on my development machine, but as soon as I use it in production environment, I get the same error... super weird stuff.

**Comment by marco-c** [comment]
@gauntface do you have any idea what's going on here? Given that it's intermittently failing, it might be a FCM problem.

Are you all seeing this problem intermittently, or is someone reproducing it all the time? Does it fail for specific messages?

**Comment by gauntface** [comment]
Looking at comment #3 are you generating new vapid tokens every time you run the script?

Chrome is much more restrictive on it's checks for VAPID headers i.e. the public VAPID key used for the application server key MUST be used with the matching private key to generate the JWT.

Is anyone able to share an example project/ of this happening?

**Comment by spock123** [comment]
@marco-c  I started implementing this 4 days ago and never got it working -so far. And - I only generated the VAPID keys once ( I know your comment wasn't for me, just to clear that up).

Thanks again guys

**Comment by gauntface** [comment]
@spock123 are you certain that the VAPID keys are the correct pair and the public application server key is passed into the subscribe() call?

Is there an example of your code? Or an example that produces the error?

**Comment by gauntface** [comment]
Can I also get @spock123 @vkammerer @fix2015 @p-nowak to just let me know the following:

1.) Is the error intermittent / constant for all subscribtions?
2.) Is the error intermittent / constant for a specific subscription that you are testing against but others are ok?
3.) What platform are you testing on? (Linux, OS X, Windows) Have you tried on other platforms? Same result?
4.) What version of node are you using.

Cheers,
Matt

**Comment by vkammerer** [comment]
I have been testing it mostly on my own laptop:
- MacOs 10.12.1
- Chrome 54
- Node 6.9.1

So I can't say if it would have behaved differently with a different subscription.

This error showed up constantly the two last weekends for a few hours in the morning, and then the service started working again.
I have posted details of the error on https://github.com/web-push-libs/web-push/issues/237, but I can repost here for convenience purposes: "

I am sending a request with an endpoint to
```
https://fcm.googleapis.com/fcm/send/ENDPOINT
```

I tried regenerating keys, unregistering workers and clearing the cache to get a whole new fresh start, but it just won't do, and I still get the following error returned by `webpush.sendNotification()`

```
{ WebPushError: Received unexpected response code
    at IncomingMessage.<anonymous> (/Users/vincentkammerer/Web/vkammerer/communauto/automobile-superpowers/node_modules/web-push/src/web-push-lib.js:274:20)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
  name: 'WebPushError',
  message: 'Received unexpected response code',
  statusCode: 400,
  headers:
   { 'content-type': 'text/html; charset=UTF-8',
     date: 'Sun, 13 Nov 2016 16:56:51 GMT',
     expires: 'Sun, 13 Nov 2016 16:56:51 GMT',
     'cache-control': 'private, max-age=0',
     'x-content-type-options': 'nosniff',
     'x-frame-options': 'SAMEORIGIN',
     'x-xss-protection': '1; mode=block',
     server: 'GSE',
     'alt-svc': 'quic=":443"; ma=2592000; v="36,35,34"',
     'accept-ranges': 'none',
     vary: 'Accept-Encoding',
     connection: 'close' },
  body: '<HTML>\n<HEAD>\n<TITLE>UnauthorizedRegistration</TITLE>\n</HEAD>\n<BODY BGCOLOR="#FFFFFF" TEXT="#000000">\n<H1>UnauthorizedRegistration</H1>\n<H2>Error 400</H2>\n</BODY>\n</HTML>\n' }
```

**Comment by gauntface** [comment]
@vkammerer Sorry when you say "the service started working again", what exactly does that mean?

What I'm trying to figure out from all of the above is:

1. Was a PushSubscription failing for a period of time and then started working?
2. Was it that your example site / service wasn't able to send notifications for any subscriptions you received from Chrome and then suddenly those same subscriptions started sending?
3. Was it that your example site / service wasn't able to send notifications for any subscriptions you received from Chrome and then at some point new subscriptions started sending?

**Comment by vkammerer** [comment]
@gauntface sorry for being unclear in my description, I'll try to be more explicit :)

My node server running locally was trying to send push notifications with `webpush.sendNotification()` using { endpoint, keys } data sent by my browser:
- between about 10am and 12pm, it would not work and I would get the error here above.
- using the exact same { endpoint, keys } data sent by the browser, and with no other change on my end, it started working again the whole afternoon.

Am I answering your question?

**Comment by vkammerer** [comment]
I should maybe mention that, on each page load, my browser gets subscription data that will be sent  with subsequent API calls to keep the server up to date.
The subscription, from what I understand, must be the same when it already exists:
https://github.com/vkammerer/automobile-superpowers/blob/master/public/cached/scripts/sw.js#L10

**Comment by gauntface** [comment]
Thats perfect, thank you @vkammerer :)

**Comment by spock123** [comment]
@gauntface  sorry for my non-answer, I've been to the gym and had a meeting :)  - I'll try and look into it tonight now (I'm in Europe) and provide some better info..

Thanks again :)

**Comment by spock123** [comment]
Ok I'm back.. 

Guys, I don't know what to say.. I haven't changed a single line of code.... 
Still, it appears to work for me in Chrome now, at least in localhost (osx, chrome 54).

I'll try and do a test in production to see if it works there as well... 

I'm confused... could it really be the fcm endpoint that is THAT unstable?

**Comment by gauntface** [comment]
@spock123 are you changing anything in terms of the VAPID keys used between testing and production? Or changing the VAPID subject?

The selenium tests for this library are running and working, they do full end-to-end tests of Chrome and Firefox.

I've raised an issue with the FCM team to see if there is anymore info that can be shared in the error message to try and make it easier to decipher what the problem is.

**Comment by spock123** [comment]
@gauntface  No, nothing. Same keys in devel and production (for the very reason of not messing with the setup).

Also, Firefox works in production - both Firefox on Android, and Firefox on OsX.

But: after some Google, you guys think this could be the issue? It regards Chrome 54:

https://bugs.chromium.org/p/chromium/issues/detail?id=623062#c37

**Comment by gauntface** [comment]
If that was the issue, you wouldn't have a subscription to send to ;) BUT you can rule out an issue with Chrome 54 by testing on Chrome Beta (55) or Chrome Dev / Canary (56)

**Comment by spock123** [comment]
Anyone on Windows7 with Chrome? You could test a deployment here:

https://osbfood.rye-jeppesen.dk

Go to settings -> click "Notify me" and then click the appearing link beneath...  

I'll try and install Chrome Beta for OSX..

**Comment by p-nowak** [comment]
@gauntface 

I found one of my other linux servers where it's working, but I made tests few minutes ago and problem still occurs, so I'm able to answer your questions:

> 1.) Is the error intermittent / constant for all subscribtions?
> 2.) Is the error intermittent / constant for a specific subscription that you are testing against but others are ok?

I tested it with only few push subscriptions , but it always worked for none or for all subscriptions

> 3.) What platform are you testing on? (Linux, OS X, Windows) Have you tried on other platforms? Same result?

Mac OS X - 3 different machines worked properly
Linux - 2 different servers with Debian 8 had issues. Currently I had it working on ubuntu server

> 4.) What version of node are you using.

All tests was made with node 6.9.1.

**Comment by p-nowak** [comment]
It looks like that, after analysis of encryption stuff I finally found a problem.

Problem happens when clock on system is set forward (even few sec). After syncing with ntp everything is working, so please try to check time on your system.

Tested on few platform and it worked, so no problem on FCM side 😄

**Comment by spock123** [comment]
@p-nowak  I tried updating using ntpdate but still getting same issue..

I should point out that my node backend runs inside a Docker container (based on node 6:6)

**Comment by p-nowak** [comment]
@spock123 
You can try to ensure that on docker environment you have correct time or for tests you can even try to set the time a few minutes backward. 

For me it resolved all problems and also tests shown that for example on Mac when I set time forward, it also stopped working with the same error.

But maybe you have different kind of problem.

**Comment by spock123** [comment]
@p-nowak  Omfg it worked!

Earlier I used ntpdate to set a precise date/time, but now I tried setting the time 5 minutes earlier than "realtime"... and now it works! 

Dafuq, Google?

Thank you so much guys, I cannot express my gratitude enough for you taking the time to look into this.. I'm sorry this isn't really to do with this excellent library but I didn't know where else to go.

THANK YOU

**Comment by marco-c** [comment]
I wonder if we can throw an exception when the clock is set in the past.

**Comment by marco-c** [comment]
Perhaps we can add an explanation to the `message` property of the `WebPushError` when the status code is 400.

**Comment by p-nowak** [comment]
@marco-c 

The problem is that status 400 means also that you can have incorrect push endpoint. Detecting from node side if user had clock ahead, could be hard :) 

Maybe it will have bigger sense to add such info in readme

**Comment by marco-c** [comment]
> The problem is that status 400 means also, that you can have incorrect (eg unregistered) push endpoint.

I wasn't suggesting to replace the message altogether, but just adding some text to explain what one possible problem might be.

E.g. instead of the current `Received unexpected response code`, we can use `Received unexpected response code. One of the possible problems is that the clock of your system is not synchronized`.

> Detecting from node side if user had clock ahead, could be hard :)

I guess we could use NTP.

> Maybe it will have bigger sense to add such info in readme

This is definitely the first thing we should do.

**Comment by marco-c** [comment]
I guess it would be better if FCM improved its error message, so we wouldn't have to implement any check or additional error message.

**Comment by gauntface** [comment]
thanks for the investigation (and sorry for not responding sooner - holiday in US).

I asked the FCM team to look into better error messages last week because this feedback cycle is painful for everyone.

Also reduced the validity of the VAPID expiration from 24hours (which is the max) to 12 hours. It sounds like this isn't the issue many were hitting because we are using Date.now() but it may help some.

**Comment by p-nowak** [comment]
I tested such change on my test env and it looks like that setting expiration to 12 hours is solving problem, for systems where clock is set forward (less than 12 hours forward, I guess)

**Comment by gauntface** [comment]
@p-nowak thank you for letting me know and glad that it helpful.

It's still not an ideal solution, but sounds like it'll help some.

**Comment by dev-mraj** [comment]
its still there in chrome 54, have tried changes with time. but not working yet.

**Comment by johndodev** [comment]
I had the error, and resolve it with unsubscribe then resubscribe in my web app. I'm still in dev so I don't have any production issues.

**Comment by JakeCallery** [comment]
Same issue here.
Seems to work fine in Firefox (currently 50.1.0 / also worked in 47.0.2)with the mozilla endpoint:
https://updates.push.services.mozilla.com/wpush/v2/

Does not work in chrome (55.0.2883.87 (Official Build) m (32-bit)) with Firebase endpoint:
https://fcm.googleapis.com/fcm/send/

**Comment by deanhume** [comment]
Not sure if this helps anyone, but @johndodev suggestion was the solution for me. I subscribed without a GCM sender id in my manifest file - but once I added it to test Chrome, I kept on getting these errors. 

Once I unsubscribed and then resubscribed all worked - no code change. Hope it helps!

**Comment by gauntface** [comment]
I'm going to close this issue for now as it looks like for most people either the time change fixed the issue and / or the problem was related to some other issue.

I'll continue to nag the Chrome / FCM team to help make this easier to diagnose and fix this error - I agree it's incredibly unhelpful error message :(

**Comment by oninross** [comment]
Can we open this issue again?  It is working fine in FF but not in Chrome.  I tried deploying it but it doesnt work at all even for FF.  Any clues what is happening?  Same error as the other developers here.  See screenshot below.  Not sure now if I am missing something or something is really wrong with the script

![screenshot_020617_025553_pm](https://cloud.githubusercontent.com/assets/2233813/22634292/62ebc8e2-ec7c-11e6-91e8-794d22d2d890.jpg)

**Comment by gauntface** [comment]
@oninross your problem looks like it's a GCM error.

In your web app if you don't set an `applicationServerKey` then Chrome will fallback to looking for a 'gcm_sender_id' (or chrome version is old). In your error, it's a GCM subscription (You can tell by the URL which is android.googleapis.com) and you don't set a GCM API key which you can tell by the first error message:

> Attempt to send push notification to GCM endpoint, undefined

The fix is to call setGCMApiKey().

https://github.com/web-push-libs/web-push#setgcmapikeyapikey

@marco-c maybe we should point to the correct function to solve this problem.

**Comment by 4DallasC** [comment]
I got rid of the UnauthorizedRegistration error from Chrome by using a '; ' in between the Crypto-Key headers. I originally had a ',' and it gave the error. Then a ';', it still gave the error. But when 1 added the space '; ', it worked.

**Comment by gauntface** [comment]
@4DallasC did you make the change in the web-push module or somewhere else?

**Comment by 4DallasC** [comment]
@gauntface I've re-written the php push library. I only write this response to tell people that Chrome will respond with this error is there are any formatting issues with the headers. Hopefully that'll give people a clue to what the problem is.

**Comment by marco-c** [comment]
In theory, according to the specs it should be a comma, but we moved to a semicolon (see #182 and the discussion in #180).
It's really strange that `; ` works for you instead of `;`.
